### PR TITLE
Refactor the /Qspectre cmake code

### PR DESCRIPTION
**Pull request by snnn**

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -144,16 +144,27 @@ if (MSVC)
...
   SET (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /wd4100")
   if (NOT onnxruntime_USE_CUDA)
     SET (CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /GL")
     SET (CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /GL")
   endif()
-  if (onnxruntime_ENABLE_QSPECTRE)
-    SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /Qspectre /guard:cf /Zi")
-    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Qspectre /guard:cf /Zi")
-    SET(CMAKE_EXE_LINKER_FLAGS  "/guard:cf /DYNAMICBASE")
+  check_cxx_compiler_flag(-Qspectre HAS_QSPECTRE)
```
Comment (jignparm) [inline cmake/CMakeLists.txt]: Does this also check that the toolset is installed? I think compiler switch should be present -- not sure it this checks if the spectre libs are installed. Worth double checking.

Comment (snnn) [reply cmake/CMakeLists.txt]: It checked it.  I verified.

File: cmake/CMakeLists.txt [hunk 2]
```diff
@@ -150,10 +150,19 @@ if (MSVC)
...
+  if (HAS_QSPECTRE)
+    SET(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /Qspectre")
+    SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} /Qspectre")
+  endif()
+  SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} /DYNAMICBASE")
```
Comment (pranavsharma) [inline cmake/CMakeLists.txt]: should this be outside the if block?

Comment (snnn) [reply cmake/CMakeLists.txt]: You mean 'DYNAMICBASE' ?

Comment (pranavsharma) [reply cmake/CMakeLists.txt]: yeah, previously it was set only when the spectre flag was enabled.

Comment (snnn) [reply cmake/CMakeLists.txt]: But it's not related to /Qspectre. It's a totally separated thing. It's an old flag.

Comment (snnn) [reply cmake/CMakeLists.txt]: By default, /DYNAMICBASE is enabled on all VC projects.

Comment (pranavsharma) [reply cmake/CMakeLists.txt]: I see. If it's the default, do we still need it?

File: cmake/CMakeLists.txt [hunk 3]
```diff
@@ -150,10 +150,19 @@ if (MSVC)
...
+  check_cxx_compiler_flag(-guard:cf HAS_GUARD_CF)
+  if (HAS_GUARD_CF)
+    SET(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} /guard:cf")
+    SET(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} /guard:cf")
+    SET(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} /guard:cf")
+    SET(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} /guard:cf")
+    SET(CMAKE_EXE_LINKER_FLAGS  "${CMAKE_EXE_LINKER_FLAGS} /guard:cf")
```
Comment (jignparm) [inline cmake/CMakeLists.txt]: Good change -- **/guard:cf** is not compatible with **/Zi**, so it should only be set separately only for Release and RelWithDebInfo configs.

Comment (snnn) [reply cmake/CMakeLists.txt]: I guess you are saying /ZI, not /Zi ?

Comment (jignparm) [reply cmake/CMakeLists.txt]: Yes, meant  **/ZI** ... as mentioned here to https://docs.microsoft.com/en-us/cpp/build/reference/guard-enable-control-flow-guard?view=vs-2017.

**Pull request merged**